controlpanel/api/cluster.py

```diff
@@ -346,3 +356,120 @@ def get_repository(user, repo_name):
         return None
+
+
+class ToolDeploymentError(Exception):
+    pass
+
+
+def deploy_tool(tool, user, **kwargs):
+    values = {
+        'username': user.username.lower(),
+        'Username': user.username.lower(),  # XXX backwards compatibility
+        'aws.iamRole': user.iam_role_name,
+        'toolsDomain': settings.TOOLS_DOMAIN,
+    }
+
+    # generate per-deployment secrets
+    for key, value in tool.values.items():
+        if value == '<SECRET_TOKEN>':
+            tool.values[key] = secrets.token_hex(32)
+
+    values.update(tool.values)
+    values.update(kwargs)
+    values = ','.join(f'{key}={val}' for key, val in values.items())
+
+    try:
+        return helm.upgrade_release(
+            f'{tool.chart_name}-{user.slug}',
+            f'mojanalytics/{tool.chart_name}',  # XXX assumes repo name
+            f'--version', tool.version,
+            f'--namespace', user.k8s_namespace,
+            f'--set', values,
+        )
+
+    except HelmError as error:
+        raise ToolDeploymentError(error)
+
+
+def list_tool_deployments(user, search_name=None, search_version=None):
+    deployments = []
+    k8s = KubernetesClient()
+    results = k8s.AppsV1Api.list_namespaced_deployment(user.k8s_namespace)
+    for deployment in results.items:
+        app_name = deployment.metadata.labels["app"]
+        _, version = deployment.metadata.labels["chart"].rsplit("-", 1)
+        if search_name and search_name not in app_name:
+            continue
+        if search_version and not version.startswith(search_version):
+            continue
+        deployments.append(deployment)
+    return deployments
+
+
+def get_tool_deployment(tool_deployment):
+    deployments = list_tool_deployments(
+        tool_deployment.user,
+        search_name=tool_deployment.tool.chart_name,
+        search_version=tool_deployment.tool.version,
+    )
+
+    if not deployments:
+        raise ObjectDoesNotExist(tool_deployment)
+
+    if len(deployments) > 1:
+        log.warning(f"Multiple matches for {tool_deployment!r} found")
+        raise MultipleObjectsReturned(tool_deployment)
+
+    return deployments[0]
+
+
+def delete_tool_deployment(tool_deployment):
+    deployment = get_tool_deployment(tool_deployment)
+    helm.delete(
+        deployment.metadata.name,
+        f"--namespace={tool_deployment.user.k8s_namespace}",
+    )
+
+
+def get_tool_deployment_status(tool_deployment):
+    try:
+        deployment = get_tool_deployment(tool_deployment)
+
+    except ObjectDoesNotExist:
+        log.warning(f"{tool_deployment} not found")
+        return TOOL_NOT_DEPLOYED
+
+    except MultipleObjectsReturned:
+        log.warning(f"Multiple objects returned for {tool_deployment}")
+        return TOOL_STATUS_UNKNOWN
+
+    conditions = {
+        condition.type: condition
+        for condition in deployment.status.conditions
+    }
+
+    if 'Available' in conditions:
+        if conditions['Available'].status == 'True':
```

Comment:
The status of k8s deployments statuses (no pun intended) is a bit of a mess.

If it helps, I've noticed that I got `Unknown` status while one of the two containers in RStudio was still starting, `type` was `"Available"` but `status` was `"False"` (what a mess!):

```YAML
status:
  conditions:
  - lastTransitionTime: "2019-09-06T14:11:01Z"
    lastUpdateTime: "2019-09-06T14:11:01Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
```

(but bear in mind that I think my RStudio is currently broken so I don't know how typical that would be)

Maybe we could have a "Starting" status? I have no idea what the k8s condition would be if the pod doesn't start propertly then, would we stuck displaying "Starting"?

For instance, my RStudio's pod is now stuck in `CrashLoopBackOff` status but the `Deployment`'s condition is still as above 😕

Reply:
I see that you got rid of the code which was checking the pods' `statuses`|`states` - I haven't looked at that properly yet so not sure if we're missing some edge case we had before but this seems so much cleaner 🌟 😍

Reply:
I don't think having a `Deploying` *and* `Starting` status is worthwhile - in the (hopefully) normal case, you would see `Starting` only very briefly between `Deploying` and `Ready`.

Are you saying that you want more code to figure out what is the reason for an `Unknown` status? I think that may not be worth the effort and I don't want to show the reason message to users, but maybe we should monitor the frequency of `Unknown` statuses to decide if it is worth spending more time on it - I'll add a `log.warning`.

Reply:
Re "deploying vs. starting" these days some tools (at least RStudio) have slow init containers but we may get away with the simple "Deploying" and iterate if necessary 👍 

The less code the happier I am you know. Hopefully that "Unknown" status is an edge case and infrequent.